drivers/hcsr04/hcsr04.c

```diff
@@ -0,0 +1,177 @@
+    dev->was_triggered = 0;
+    dev->pre_trig_t_us = xtimer_now_usec();
+    dev->pre_meas_t_us = xtimer_now_usec();
+    hcsr04_set_temp(dev, params->temperature);
+
+    if (gpio_init(dev->params.trigger_pin, GPIO_OUT)) {
+        return -EIO;
+    }
+
+    if (gpio_init_int(dev->params.echo_pin, GPIO_IN, GPIO_BOTH, hcsr04_int_callback, dev)) {
+        return -EIO;
+    }
+
+    // necessary in order for the next trigger not to fail due to time constraint 
+    xtimer_msleep(30);
+
+    return 0;
+}
+
+uint32_t hcsr04_get_max_response_time(hcsr04_t *dev) {
+    return 2000 * HCSR04_MAX_DISTANCE_MM / dev->sound_speed;
+}
+
+int hcsr04_trigger(hcsr04_t *dev)
+{
+    uint32_t now = xtimer_now_usec();
+    uint16_t min_response_time = hcsr04_get_max_response_time(dev);
+    
+    if ((now - dev->pre_trig_t_us) < min_response_time) {
+        return -EFREQ;
+    }
+    
+    dev->was_triggered = 1;
+    dev->pre_trig_t_us = now;
+    gpio_set(dev->params.trigger_pin);
+    xtimer_usleep(TRIGGER_TIME);
+    gpio_clear(dev->params.trigger_pin);
+    
+    mutex_unlock(&dev->lock);
+    mutex_lock(&dev->lock);
+
+    return 0;
+}
+
+int hcsr04_read(hcsr04_t *dev, uint16_t *distance)
+{
+    if (!dev->was_triggered) {
+        *distance = 0;
+        return -ENTR;
+    }
+
+    dev->was_triggered = 0;
+
+    // try to lock it again, should be released when it is unlocked or cause of timeout
+    uint16_t min_response_time = hcsr04_get_max_response_time(dev);
+    int status = xtimer_mutex_lock_timeout(&dev->lock, min_response_time);
+    
+    // unlock it anyway
+    mutex_unlock(&dev->lock);
+
+    if (status == -1) {
+        *distance = 0;
+        return -ECON;
+    }
+    
+    if (dev->distance < HCSR04_MIN_DISTANCE_MM) {
+        *distance = HCSR04_MIN_DISTANCE_MM;
+    } else if (dev->distance > HCSR04_MAX_DISTANCE_MM) {
+        *distance = HCSR04_MAX_DISTANCE_MM;
+    } else {
+        *distance = dev->distance;
+    }    
+
+    return 0;
+}
+
+uint16_t hcsr04_temp_to_sound_speed(uint16_t temp) 
+{
+    return COEFFICIENT * sqrt(REF_TEMPERATURE + temp / 1000);
+}
+
+void hcsr04_set_temp(hcsr04_t *dev, uint16_t new_temp)
+{
+    dev->params.temperature = new_temp;
+    dev->sound_speed = hcsr04_temp_to_sound_speed(new_temp);
+}
+
+int hcsr04_get_distance(hcsr04_t *dev, uint16_t *distance) 
+{
+    *distance = 0;
+
+    int status = hcsr04_trigger(dev);
+    if (status < 0) {
+        return status;
+    }
+
+    status = hcsr04_read(dev, distance);
+    if (status < 0) {
+        return status; 
+    }
+        
+    return 0;
+}
+
+void hcsr04_deinit(hcsr04_t *dev) 
+{
+    gpio_clear(dev->params.trigger_pin);
+    gpio_irq_disable(dev->params.echo_pin);
+
+    dev->params.temperature = 0;
+    dev->params.trigger_pin = 0;
+    dev->params.echo_pin = 0;
+
+    mutex_unlock(&dev->lock);
+    dev->distance = 0;
+    dev->was_triggered = 0;
+    dev->sound_speed = 0;
+    dev->pre_trig_t_us = 0;
+    dev->pre_meas_t_us = 0;
```

Comment:
```suggestion
    dev->params.temperature = 0;
    dev->params.trigger_pin = 0;
    dev->params.echo_pin = 0;

    mutex_unlock(&dev->lock);
    dev->distance = 0;
    dev->was_triggered = 0;
    dev->sound_speed = 0;
    dev->pre_trig_t_us = 0;
    dev->pre_meas_t_us = 0;
```

I think only the GPIO stuff could be left as if the memory of the device descriptor gets unused/deallocated it wont matter that much if it's all zero